src/analyses/apron/apronPriv.apron.ml

```diff
@@ -983,22 +983,37 @@ struct
       let l' = L.add lm apr_side l in
       {apr = apr_local; priv = (w',LMust.add lm lmust,l')}
 
-  let thread_join (ask:Q.ask) getg exp (st: apron_components_t) =
+  let thread_join ?(force=false) (ask:Q.ask) getg exp (st: apron_components_t) =
     let w,lmust,l = st.priv in
     let tids = ask.f (Q.EvalThread exp) in
-    if ConcDomain.ThreadSet.is_top tids then
-      st (* TODO: why needed? *)
+    if force then (
+      if ConcDomain.ThreadSet.is_top tids then
+        st (* don't consider anything more joined, matches threadJoins analysis *)
+      else (
+        (* fold throws if the thread set is top *)
+        let (lmust', l') = ConcDomain.ThreadSet.fold (fun tid (lmust, l) ->
+            let lmust',l' = G.thread (getg (V.thread tid)) in
+            (LMust.union lmust' lmust, L.join l l')
+          ) tids (lmust, l)
+        in
+        {st with priv = (w, lmust', l')}
+      )
+    )
```

Comment:
A perhaps surprising effect here is that the `__goblint_assume_join(...)` annotation may destroy precision here, if things have actually been must-joined before and have then definitely been overwritten locally before (I added an example).

I think here it should be sufficient to iterate over those not already consider must-joined anyway?

Reply:
Wouldn't the same problem also happen if you repeat `pthread_join(id);` where the thread ID is unique and can normally be must-joined?

Reply:
Yes, but joining things twice is UB, so I don't really think it matters in that case that we become imprecise (we should in fact produce a gigantic warning if anyone does this!)

I would understand this `__goblint_assume_join(...)` not as something that one would put into the program instead of the actual join, but something one additionally can put to give Goblint the hint that the join must have happened. In this case it is surprising if Goblint suddenly becomes less precise because of it. Or did I misunderstand it?

Reply:
> Yes, but joining things twice is UB, so I don't really think it matters in that case that we become imprecise (we should in fact produce a gigantic warning if anyone does this!)

That would be part of #176.


> I would understand this `__goblint_assume_join(...)` not as something that one would put into the program instead of the actual join, but something one **additionally** can put to give Goblint the hint that the join must have happened. In this case it is surprising if Goblint suddenly becomes less precise because of it. Or did I misunderstand it?

You are correct, that it's an additional thing. I can add an extra check about already must-joined threads there to avoid this possibility.
Although actually you would only ever add the assume join if at the actual join Goblint couldn't must-join it, so the assume join would be the only one doing its must join.

Reply:
I think that would be nice. I can imagine situations where one desperately tries things until something works and I could see it happening that someone inserts such an annotation ever for a thread where we find it is must-joined.

Reply:
Apparently the annotation doesn't help on chrony any more for some reason. I'll look into it.

Reply:
Due to #716 in chrony the assume-joined thread ID turns out to be top, so by making it safer and ignoring that, the chrony story collapses again. I don't think we'll be able to fix every single issue that causes unknown thread IDs to occur in chrony.

Maybe a workable alternative is to still assume-join everything with unknown thread ID, but emit an unsoundness message when that happens? Then it would be apparent, when you should do the equivalent thing and remove the annotation (which would have no effect anyway), but allow the user to override our super imprecision if they really know what they're doing.

Reply:
As discussed in GobCon, I changed it back to allow assume-join of unknown thread ID to consider everything joined, but a warning is printed. Also when any thread is then created, the must-joined set is made empty and another warning is printed.
This now works for the chrony story and hopefully is reasonable enough behavior.